_appmap/testing_framework.py

```diff
@@ -161,3 +161,4 @@ def failure_location(exn: Exception) -> str:
         path = root_relative_path(frame.filename)
         if not PurePath(path).is_absolute():
             return f"{path}:{frame.lineno}"
+    return "<unknown>"  # should never happen?
```

Comment:
I'm not sure. Are there actually cases where all the frames in a stack trace, even those for app code, have absolute paths? If there are, then it would probably be better to type the return value as `Optional[str]`, check for `None` in the caller`, and leave it out of the metadata if it's `None`. If there aren't, then yes, this should do something more drastic.

Reply:
We're making the paths relative to root with `root_relative_path`, so as long as the exception is thrown by the application code we should always end up having at least one — but you're right, we might get an unrelated exception. Maybe in this case we should just return the location of the first stack frame (even if it's absolute)?